[lua] Add multireturn behavior handling for lua externs

jdonaldson opened the pull request:
# DON'T MERGE THIS YET, STILL UNDER DISCUSSION

This is the third version of multireturn behavior.  This version detects
when a variable is assigned from the result of a Meta.Multireturn method
call.  In this case, the Meta.MultiReturn meta is added to the v_meta
field of the local variable, along with a local mapping of the field
names to the temp name aliases.

This method is in contrast to the previous attempt, which used a special
v_multi field on the variable itself.

jdonaldson commented on src/generators/genlua.ml:
```diff
@@ -278,7 +277,56 @@ let rec is_string_type t =
 
 let is_string_expr e = is_string_type e.etype
-(* /from genphp *)
+
```
This block is all the helper functions that deal with extracting values out of the v_meta fields for a local variable, or building or finding an index.

jdonaldson commented on src/generators/genlua.ml:
```diff
@@ -477,6 +525,10 @@ and gen_expr ?(local=true) ctx e = begin
 		spr ctx "]";
 	| TBinop (op,e1,e2) ->
 		gen_tbinop ctx op e1 e2;
+	| TField ({eexpr = TCall _} as e, f) when is_multireturn ctx e ->
```
If we're immediately extracting a multireturn value from a method call that generates one, we use the "select" method since it doesn't require creating any temp variables.

jdonaldson commented on src/generators/genlua.ml:
```diff
@@ -512,8 +564,14 @@ and gen_expr ?(local=true) ctx e = begin
+	| TField ({eexpr = TLocal tv} as e, f) when is_multireturn ctx e ->
+		(match f with
+		| FInstance(_,_,cf)->
+			let tmpname = multi_value_tmp_lookup tv.v_meta cf.cf_name in
```
Here's where we look up the temporary name of a multireturn variable field, and use that instead of field access.

jdonaldson commented on src/generators/genlua.ml:
```diff
@@ -477,6 +525,10 @@ and gen_expr ?(local=true) ctx e = begin
+	| TField ({eexpr = TCall _} as e, f) when is_multireturn ctx e ->
+		print ctx "_G.select(%i, " (multi_value_idx_lookup_expr ctx e (field_name f));
+		gen_value ctx e;
+		spr ctx ")";
```
If we're immediately accessing a multireturn field value that is the result of a method call, then just use lua's select() method.  This avoids having to create any temporary variables.

jdonaldson commented on src/generators/genlua.ml:
```diff
@@ -578,6 +636,15 @@ and gen_expr ?(local=true) ctx e = begin
+				    v.v_name <- String.concat ", " local_tmp_names;
+				    print ctx "local %s =" v.v_name;
+				    gen_value ctx e;
+				    semicolon ctx;
```
If we do need to create temporary variables, we create some metadata on the TVar that keeps a map between the original field name, and the generated temp name.  This is done by creating a Meta.MultiReturn metadata on the TVar, and then creating an EOBjectDecl that contains a series of name -> EConst(String("name")) tuples.

jdonaldson commented on src/generators/genlua.ml:
```diff
@@ -1630,6 +1698,22 @@ let generate_require ctx path meta =
+    | {cl_super = Some(csup,_)} when Meta.has Meta.MultiReturn csup.cl_meta ->
+		error "Cannot extend a MultiReturn" c.cl_pos
+    | _ -> ()
+
```
Here's the checks we need to run for the multireturn.  @nadako suggested we move these into the typer, but I hadn't heard any other feedback.

jdonaldson commented on src/generators/genlua.ml:
```diff
@@ -1649,7 +1733,9 @@ let generate_type ctx = function
-			| _ -> generate_package_create ctx c.cl_path)
+			| _ -> generate_package_create ctx c.cl_path);
+
+		check_multireturn ctx c;
```
Check for correct multireturn behavior inside of generate_type (for now?).

jdonaldson commented:
Here's some example usage:

```haxe
class Main {

    public static function main(){
        trace(find("foo bar", "foo").begin + " is the value for k"); // we can access the fields of the return immediately
        var k = find("foo bar", "foo"); // we can assign to a new variable, which has special semantics.
        var l:Dynamic = k.begin; // we can assign out of the multireturn type to a new variable
        trace(l + " is the value for l"); 
        trace(k.end + " is the value for k.end");

        // the multireturn reference is just a list of variables, 
        // so it doesn't print correctly... yet.
        trace(k + " is the value for k"); 

        return k; // but, we can return multireturn types
    }


    // Lua's string "find" method returns two values, it's a typical 
    // multireturn value case we want to cover.  Here we're aliasing
    // it inside a helper method that declares the result to be a multireturn
    // "StringFind" type.
    static function find(str1:String, str2:String) : StringFind {
        return untyped lua.NativeStringTools.find(str1, str2);
    }

}

@:multiReturn extern class StringFind {
    var begin : Int;
    var end   : Int;
}
```

Here's the key output of the resulting lua:

```lua
Main.new = {}
Main.__name__ = true
Main.main = function() 
  haxe.Log.trace(_G.select(1, Main.find("foo bar","foo")) .. " is the value for k",_hx_o({__fields__={fileName=true,lineNumber=true,className=true,methodName=true},fileName="Main.hx",lineNumber=4,className="Main",methodName="main"}));
  local _hx_1_begin, _hx_2_end =Main.find("foo bar","foo");
  haxe.Log.trace(Std.string(_hx_1_begin) .. " is the value for l",_hx_o({__fields__={fileName=true,lineNumber=true,className=true,methodName=true},fileName="Main.hx",lineNumber=7,className="Main",methodName="main"}));
  haxe.Log.trace(_hx_2_end .. " is the value for k.end",_hx_o({__fields__={fileName=true,lineNumber=true,className=true,methodName=true},fileName="Main.hx",lineNumber=8,className="Main",methodName="main"}));
  haxe.Log.trace(Std.string(_hx_1_begin, _hx_2_end) .. " is the value for k",_hx_o({__fields__={fileName=true,lineNumber=true,className=true,methodName=true},fileName="Main.hx",lineNumber=12,className="Main",methodName="main"}));
  do return _hx_1_begin, _hx_2_end end;
end
Main.find = function(str1,str2) 
  do return _G.string.find(str1,str2) end;
end
```

nadako commented:
Initial thoughts (copied from the Slack):
 * I still think multi-return values should be autoboxed into an object `{begin=_hx_1_begin, end=_hx_2_end}` when unified with Dynamic or a structure type, or when passed as a value (e.g as an argument or field value). That would follow the principle of least surprise: it would "just work", also reading the haxe code user would expect an object created anyway (since it's a class), so to me this is more of an optimization (though a very important one!).
 * The `check_multireturn` should be done somewhere before the DCE so unused classes are still checked.
 * It's `@:multiReturn`, not `@:MultiReturn` :)

I haven't yet looked at the actual code, but I'm pretty sure it's fine :)

Simn commented:
Running `check_multireturn` before DCE would indeed be better, but it's not *that* important and can always be changed later. Other than that I'm fine with the PR.

jdonaldson commented:
Thanks for the feedback.  This feature feels a bit like a square peg in a round hole, but I think it's worth supporting for externs.  The last thing I think I will add is special support for Std.string, which should enable a compatible trace and string representation of these types.  I should have it this weekend.

Moving forward, perhaps we fold this into a more general tuple or destructured assignment feature.

jdonaldson commented:
Thanks for the discussion here.  I think this is a bit messy, but a better approach has not revealed itself yet.

The last thing I think @nadako and I really wanted was an easy way to print these things out to see what the multireturn contained.  

I implemented the functionality for that, and made some more fixes. But, rather than tack it on to this review, I squashed all the changes and put them in a new pull request over here  https://github.com/HaxeFoundation/haxe/pull/5563

That pull review doesn't have unit tests yet for this feature, those are still pending.
